app/src/src_nonfreenet/org/breezyweather/sources/cwa/CwaResultConverter.kt

```diff
@@ -0,0 +1,645 @@
+    calendar.set(year, month-1, day, hour, 0, 0)
+    if (hour == 0 || hour == 12) {
+        calendar.add(Calendar.HOUR_OF_DAY, -6)
+    }
+    val result = formatter.format(calendar.time)
+    return result
+}
+
+private fun getWindDirection(direction: String?): Double? {
+    return if (direction == null) {
+        null
+    } else when (direction) {
+        "偏北風" -> 0.0
+        "東北風" -> 45.0
+        "偏東風" -> 90.0
+        "東南風" -> 135.0
+        "偏南風" -> 180.0
+        "西南風" -> 225.0
+        "偏西風" -> 270.0
+        "西北風" -> 315.0
+        else -> null
+    }
+}
+
+// Weather icon source:
+// https://opendata.cwa.gov.tw/opendatadoc/MFC/A0012-001.pdf
+private fun getWeatherCode(icon: String?): WeatherCode? {
+    return if (icon == null) {
+        null
+    } else when (icon) {
+        "01", "02" -> WeatherCode.CLEAR
+        "03", "04" -> WeatherCode.PARTLY_CLOUDY
+        "05", "06", "07" -> WeatherCode.CLOUDY
+        "08", "09", "10", "11", "12", "13", "14", "19", "20", "29", "30", "31", "32", "38", "39" -> WeatherCode.RAIN
+        "15", "16", "21", "22", "33", "34", "35", "36" -> WeatherCode.THUNDER
+        "17", "18", "41" -> WeatherCode.THUNDERSTORM
+        "23", "37", "40" -> WeatherCode.SLEET
+        "24", "25", "26", "27", "28" -> WeatherCode.FOG
+        "42" -> WeatherCode.SNOW
+        else -> null
+    }
+}
+
+private fun getAlertSeverity(headline: String): AlertSeverity {
+    return when (headline) {
+        // missing severity levels for the following because we are not sure about wording in the API JSON yet
+        // 低溫特報 (嚴寒, 非常寒冷, 寒冷), 高溫資訊 (紅燈, 橙燈, 黃燈)
+        "超大豪雨特報" -> AlertSeverity.EXTREME
+        "大豪雨特報", "海上陸上颱風警報", "陸上颱風警報", "海嘯警報" -> AlertSeverity.SEVERE
+        "豪雨特報", "海上颱風警報", "海嘯警訊" -> AlertSeverity.MODERATE
+        "熱帶性低氣壓特報", "大雨特報", "海嘯消息", "濃霧特報", "長浪即時訊息", "陸上強風特報", "海上強風特報" -> AlertSeverity.MINOR
+        else -> AlertSeverity.UNKNOWN
+    }
+}
+
+// Color source: https://www.cwa.gov.tw/V8/assets/css/main.css
+private fun getAlertColor(headline: String, severity: AlertSeverity): Int {
+    return when (headline) {
+        "陸上強風特報" -> Color.rgb(230, 229, 98)
+        "濃霧特報" -> Color.rgb(151, 240, 60)
+        else -> when (severity) {
+            AlertSeverity.EXTREME -> Color.rgb(214, 0, 204)
+            AlertSeverity.SEVERE -> Color.rgb(255, 0, 0)
+            AlertSeverity.MODERATE -> Color.rgb(255, 128, 0)
+            AlertSeverity.MINOR -> Color.rgb(255, 255, 2)
+            else -> Color.rgb(237, 146, 156)
+        }
+    }
+}
+
+// Concentrations of SO₂, NO₂, O₃ are given in ppb (and in ppm for CO).
+// We need to convert these figures to µg/m³ (and mg/m³ for CO).
+private fun ppbToUgm3(pollutant: String, ppb: Double?): Double? {
+    if (ppb == null) return null
+    val molecularMass: Double = when (pollutant) {
+        "NO2" -> 46.0055
+        "O3" -> 48.0
+        "SO2" -> 64.066
+        "CO" -> 28.01
+        else -> return null
+    }
+    // assume 25C
```

Comment:
I don't think you can assume it's always 25 °C in Taiwan

Reply:
I will add temperature as a parameter to this function, so the concentration calculation can be more precise based on the reported ambient temperature.

I remember that the formula used assumes 1 atmospheric pressure as well. I'll see if I can find a more detailed formula that allows plugging in the air pressure. If so, I'll add that as a parameter also.

Reply:
I figured out how one of the constants in the original formula was derived from the molar gas constant but applied at 1 atm (1013.25 hPa). I have update the function to take both temperature and air pressure into account.

I'm not going to make `convertSecondary` loop through the labyrinthine API response to get the temperature and air pressure. So if null values are passed, the temperature and air pressure will be assumed at 25°C and 1 atm (1013.25 hPa) respectively, which are [commonly used](https://en.wikipedia.org/wiki/Useful_conversions_and_formulas_for_air_dispersion_modeling) parameters.